server.py

```diff
@@ -1,36 +1,376 @@
+        if data.get("housing") in coach.housing:
+            match_score += 10
+        if data.get("gender") == coach.gender:
+            match_score += 5
+        if "birth_year" in data:
+            match_score += max(0, 10 - abs(data["birth_year"] - coach.birth_year))
+        coach_matches.append((match_score, coach))
+
+    return sorted(coach_matches, reverse=True)
+
+
+# ------------------------------------------------------------------------------
+# Routes
+# ------------------------------------------------------------------------------
+
+
+COMMON_HEADERS = {"Access-Control-Allow-Origin": "http://localhost:3000"}
 
-    return decorated_function
 
 @app.route("/")
-@authorized
-async def index(request):
-  return await sanic.response.file("build/index.html")
+async def index(request: sanic.request.Request) -> sanic.response.HTTPResponse:
+    return await sanic.response.file("build/index.html")
+
+
+@app.route("/api/v1/coaches", methods=["GET", "POST"])
+@app.route("/api/v1/coaches/<coach_id:int>", methods=["GET", "POST", "DELETE"])
+async def api_coaches(
+    request: sanic.request.Request, coach_id: Optional[int] = None
+) -> sanic.response.HTTPResponse:
+    """HTTP API for 'coaches'."""
+    try:
+        if request.method == "GET":
+            if coach_id is None:
+                coaches = get_coaches()
+                response = sanic.response.json([c.to_json() for c in coaches])
+            else:
+                coach = get_coach(coach_id)
+                response = sanic.response.json(coach.to_json())
+        elif request.method == "POST":
+            if coach_id is None:
+                coach = create_coach(request.json)
+            else:
+                coach = edit_coach(coach_id, request.json)
+            response = sanic.response.json(coach.to_json())
+        elif request.method == "DELETE":
+            delete_coach(coach_id)
+            response = sanic.response.empty()
+        else:
+            assert False
+    except Exception:
+        logger.exception(
+            "Unexpected exception on %s route", request.path, exc_info=True
+        )
+        response = sanic.response.text("Error processing request", status=400)
+
+    response.headers = sanic.response.Header(COMMON_HEADERS)
+    return response
+
+
+@app.route("/api/v1/coach-matches", methods=["GET"])
+async def api_coach_matches(
+    request: sanic.request.Request,
+) -> sanic.response.HTTPResponse:
+    """HTTP API for 'coach-matches'."""
+    try:
+        args = {k: v[0] for k, v in request.args.items()}
+        for int_arg in ["birth_year", "need", "rights", "housing"]:
+            if int_arg in args:
+                args[int_arg] = int(args[int_arg])
+        if "languages" in args:
+            args["languages"] = _parse_languages(args["languages"])
+        coach_matches = get_coach_matches(args)
+        response = sanic.response.json(
+            [
+                {**coach.to_json(), "match_score": score}
```

Comment:
I quite like the nesting idea since it feels like 'coach' and 'match_score' should be at the same level - but then conversely the 'match_score' is arguably an attribute of the coach, so I also don't really mind!

Reply:
Yeah it's not clear which is best, I think if there were more fields it would make more sense to nest 'coach', but for now I'll leave it like this.